feat: [CI-8387]: Persist RCA in logservice db

shubham149 opened the pull request:
## Describe your changes

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/49180)
<!-- Reviewable:end -->

anurag-harness commented on product/log-service/db/mongodb/mongo_db.go:
```diff
@@ -0,0 +1,138 @@
+	if err = client.Connect(ctx); err != nil {
+		return nil, err
+	}
+	// Ping mongo server to see if it's accessible. This is a requirement for startup of TI service.
```
Stale log line

shubham149 replied:
Done

shubham149 commented:
utsall

shubham149 commented:
trigger utsall

shubham149 commented:
trigger uts

vistaarjuneja commented on product/log-service/handler/intelligence.go:
```diff
@@ -128,8 +165,46 @@ func HandleRCA(store store.Store, cfg config.Config) http.HandlerFunc {
+		report.Vote = vote
+		if err := saveToDB(ctx, id, report, mongodb); err != nil {
+			WriteInternalError(w, err)
+			logger.FromRequest(r).
```
this log should come before we WriteInternalError

vistaarjuneja commented on product/log-service/handler/intelligence.go:
```diff
@@ -128,8 +165,46 @@ func HandleRCA(store store.Store, cfg config.Config) http.HandlerFunc {
+		if err != nil {
+			WriteNotFound(w, err)
+			logger.FromRequest(r).
+				WithError(err).
```
Ditto: move this before returning back

vistaarjuneja commented on product/log-service/handler/intelligence.go:
```diff
@@ -128,8 +165,46 @@ func HandleRCA(store store.Store, cfg config.Config) http.HandlerFunc {
+	}
+}
+
 func retrieveLogRCA(ctx context.Context, endpoint, secret, provider,
```
please add comments wherever possible, eg: what is regenerate

shubham149 replied:
Done

vistaarjuneja commented:
@shubham149 please update log service README with:
i) how to connect to mongo and run migrate script for local
ii) example of curl API call for RCA

shubham149 commented:
Added readme for mongo setup for log-service. No extra envs needs to be set for running log-service with mongo in local.
I have removed /rca api from README on suggestion of @hemanthmantri since users can start calling this api in prod and it will cost us.

shubham149 commented:
trigger utsall

shubham149 commented:
trigger uts

shubham149 commented:
trigger uts

shubham149 commented:
trigger ut1

hemanthmantri commented on product/log-service/handler/intelligence.go:
```diff
@@ -89,22 +97,22 @@ Here is the logs, remember to give the response only in json format like the exa
-	vertexAIProvider = "vertexai"
-	vertexAIModel    = "text-bison"
+	genAITemperature      = 0.0
+	genAIRegenTemperature = 0.7
```
This should change with the number of regenerate call (which can be sent via API) so this should be a function of the regen number instead of a constant.

shubham149 replied:
UI won't know the regenerate call count

hemanthmantri commented on product/log-service/handler/intelligence.go:
```diff
@@ -154,17 +186,21 @@ func HandleRCA(store store.Store, cfg config.Config) http.HandlerFunc {
 			return
 		}
 
+		if err := saveToDB(ctx, id, report, mongodb); err != nil {
```
Does it do retries?

shubham149 replied:
retries are enabled on read/write by default.

```
// SetRetryWrites specifies whether supported write operations should be retried once on certain errors, such as network
// errors.
//
// Supported operations are InsertOne, UpdateOne, ReplaceOne, DeleteOne, FindOneAndDelete, FindOneAndReplace,
// FindOneAndDelete, InsertMany, and BulkWrite. Note that BulkWrite requests must not include UpdateManyModel or
// DeleteManyModel instances to be considered retryable. Unacknowledged writes will not be retried, even if this option
// is set to true.
//
// This option requires server version >= 3.6 and a replica set or sharded cluster and will be ignored for any other
// cluster type. This can also be set through the "retryWrites" URI option (e.g. "retryWrites=true"). The default is
// true.
func (c *ClientOptions) SetRetryWrites(b bool) *ClientOptions {
```

```
// SetRetryReads specifies whether supported read operations should be retried once on certain errors, such as network
// errors.
//
// Supported operations are Find, FindOne, Aggregate without a $out stage, Distinct, CountDocuments,
// EstimatedDocumentCount, Watch (for Client, Database, and Collection), ListCollections, and ListDatabases. Note that
// operations run through RunCommand are not retried.
//
// This option requires server version >= 3.6 and driver version >= 1.1.0. The default is true.
func (c *ClientOptions) SetRetryReads(b bool) *ClientOptions {
```

hemanthmantri commented on product/log-service/handler/intelligence.go:
```diff
@@ -190,9 +268,14 @@ func retrieveLogRCA(ctx context.Context, endpoint, secret, provider,
-	response, isBlocked, err := predict(ctx, client, provider, prompt)
+	temp := genAITemperature
+	if regenerate {
+		temp = genAIRegenTemperature
```
This should a function of regen count that can be fixed (and rotated back). Eg: 1,2,3,4,5,1,2,3,4,5,1 ....

shubham149 replied:
UI won't know the regenerate call count

hemanthmantri commented on product/log-service/handler/intelligence.go:
```diff
@@ -465,3 +551,45 @@ func getKeys(r *http.Request) ([]string, error) {
+		return nil, errors.New("id is required")
+	}
+
+	if mongoDB == nil {
```
This is now making it in the hot path, right?

shubham149 replied:
We won't save it to db if mongo conn is not present.

hemanthmantri commented on product/log-service/migrations/mongodb/0001_setup_database.up.json:
```diff
@@ -0,0 +1,5 @@
+[
```
Please doc the schema and fields in the README file of the log service.

shubham149 replied:
Updated

hemanthmantri reviewed (commented):
cc: @rutvijmehta-harness @anurag-harness

shubham149 commented:
trigger uts

rutvijmehta-harness commented on product/log-service/db/mongodb/mongo_db.go:
```diff
@@ -0,0 +1,143 @@
+
+func (mdb *MongoDb) FindOne(ctx context.Context, collection string, filter interface{},
+	out interface{}) error {
+	err := mdb.Database.Collection(collection).FindOne(ctx, filter).Decode(out)
```
I'd prefer to use Analytics node for read operations if there's one provisioned

rutvijmehta-harness commented on product/log-service/handler/intelligence.go:
```diff
@@ -465,3 +551,45 @@ func getKeys(r *http.Request) ([]string, error) {
+		LogID:  id,
+		Report: *report,
+	}
+	if _, err := retrieveFromDB(ctx, id, mongoDB); err != nil {
```
We can use replaceOne to insert the document if it doesn't exist: https://www.mongodb.com/docs/manual/reference/method/db.collection.replaceOne/

bot-harness commented:
This pull request has been automatically closed because it hasn't been updated in 25 days.
